VIDEO/DISPLAY: add ELCDIF support and VIDEO/DISPLAY shim layer

loicpoulain opened the pull request:
Create a shim layer to use generic video display with regular display API.
Add mcux_elcdif video driver + RK043 driver
Add camera capture and display sample

Add RK043FN02H LCD driver (video).
Add NXP MCUX ELCIF driver (video).
Add camera capture to LCD sample.
Add video to display driver (helpers).
Add support for display inheritance to video.

modules/hal/nxp/ change necessary: https://github.com/zephyrproject-rtos/hal_nxp/pull/21

Commit:
drivers: video: Add RK043FN02H driver

Video subdev driver for RK043FN02H TFT display.
Supporting RK043FN02H-CT 4.3 inch TFT 480*272 panel.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

carlescufi reviewed (approved):
LGTM

Commit:
samples: video: Add capture & display sample

This sample relies on a video capture device (e.g. pattern generator,
camera) and output device (e.g. display) to create a zerocopy capture
and display pipeline.

It has been tested on mimxrt1064-evk with rk043fn02h-ct display and
both MT9M114 camera and Software pattern generator.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

vanwinkeljan commented on drivers/video/video_display_legacy.c:
```diff
@@ -0,0 +1,175 @@
+ */
+
+/*
+ * video_display_legacy is a shim layer allowing to expose a display compatible
```
Wondering if it would not be a better idea to add a generic video_display_driver to `drivers/display` that forwards the calls to an associated video display device (a driver in `drivers/video`) in this way no extra code needs to be added to the driver in `drivers/video`.

Further I dont think it is a good idea to call it `legacy` as far as I can see the `video` API assumes it operates on full frame buffers where as the read/write function in the `display` API are designed to work with partial buffer as a lot of devices will not have enough RAM to operate with full frame buffers.

loicpoulain replied:
yes will change this misleading name

loicpoulain replied:
@vanwinkeljan , could you please review again

vanwinkeljan replied:
@loicpoulain had a quick look but should the new implementation not create a display device that has a pointer to the video device? All calls made to the de display device can then be forwarded to the video device.

loicpoulain replied:
well, for me we just have one device implementing two APIs. Adding a new dummy display device is possible but I'm not sure what it would bring... except maybe removing some display specific code from the video driver.

vanwinkeljan replied:
To me it looks like now a third API has been introduced:
1) display API
2) video API
3) display video API

I thought the intention of the shim layer was to reuse existing display code.

Or is just a DEVICE_AND_API_INIT missing in `drivers/video/video_mcux_elcdif.c`  to register the shim layer?

loicpoulain replied:
> To me it looks like now a third API has been introduced:
> 
>     1. display API
> 
>     2. video API
> 
>     3. display video API
> 
> 
> I thought the intention of the shim layer was to reuse existing display code.
> 
> Or is just a DEVICE_AND_API_INIT missing in `drivers/video/video_mcux_elcdif.c` to register the shim layer?

No still 2 APIs, It's just that video_mcux_elcdif has an option/config to also support 'display API' (via shim layer). So at the end, there is one device implementing both video and display APIs.

vanwinkeljan replied:
Problem that I currently see is that the display API assumes that a device driver implements the display API. So if you pass a `dev` to any of display API functions it casts the `driver_api` member to `struct display_driver_api` which is not valid for the proposed `video_mcux_elcdif` driver as the `driver_api` member for `video_mcux_elcdif` driver is of the type `struct video_driver_api`


https://github.com/zephyrproject-rtos/zephyr/blob/2d9032e5d09e12ed8efc1132b1666583c9689895/include/drivers/display.h#L239-L248

loicpoulain replied:
the **dev->driver_api** can be either cast as **video_driver_api** or **display_driver_api** because video_driver_api inherits display_driver_api (cf include/drivers/video.h).

https://github.com/zephyrproject-rtos/zephyr/blob/064930a565065c53f256a99bfbd88ce766e2c442/include/drivers/video.h#L233

vanwinkeljan replied:
ha that was the line I missed :)

Thx for pointing out and sorry for all the miscommunication

vanwinkeljan commented on drivers/video/video_display_legacy.c:
```diff
@@ -0,0 +1,175 @@
+		return err;
+	}
+
+	/* LVGL no double buffering aware support ? just copy previous one */
```
Bit off topic but changes could be made to the LVGL glue logic to directly support the video API

loicpoulain replied:
yes, I put that on the TODO list

dbkinder commented on samples/video/capture_and_display/README.rst:
```diff
@@ -0,0 +1,60 @@
+Requirements
+************
+
+This sample requires a video capture device (e.g. a camera).
```
... and a display.

(The mimxrt1064_evk board has a display connector, but no built-in display.)

loicpoulain replied:
fixed

dbkinder commented on samples/video/capture_and_display/README.rst:
```diff
@@ -0,0 +1,60 @@
+******
+
+On :ref:`mimxrt1064_evk`, The MT9M114 camera module should be plugged in the
+J35 camera connector. A USB cable should be connected from a host to the micro
```
... and connect the display...

loicpoulain replied:
fixed

dbkinder commented on samples/video/capture_and_display/README.rst:
```diff
@@ -0,0 +1,60 @@
+      RGBP width [640; 640; 0] height [480; 480; 0]
+      Default format: RGBP 640x480
+    Capture started
+
```
Is this output on the console I presume? And does the camera capture will show on the display?

loicpoulain replied:
Yes added description of display output

dbkinder commented on samples/video/capture_and_display/README.rst:
```diff
@@ -0,0 +1,60 @@
+
+References
+**********
+
```
If you want these items to show up under this References heading, add:
```
.. target-notes::

```
and a blank like before the first link reference.

loicpoulain replied:
Ok done, thanks

vanwinkeljan reviewed (approved):
+1 on the display related changes

vanwinkeljan commented:
@jfischer-phytec-iot could you have a look at the display related changes

jfischer-phytec-iot commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,179 @@
+		return;
+	}
+
+	printk("init\n");
```
LOG_DBG?

loicpoulain replied:
going to remove that

jfischer-phytec-iot commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,179 @@
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display_video);
```
If you do not use LOG_LEVEL in the code just use `LOG_MODULE_REGISTER(display_video, CONFIG_LOG_DEFAULT_LEVEL);` instead.

jfischer-phytec-iot replied:
also please change it to CONFIG_DISPLAY_LOG_LEVEL

loicpoulain replied:
done

jfischer-phytec-iot commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,179 @@
+ */
+
+/*
+ * display_video is a shim layer allowing to expose a display compatible
```
display API compatible interface?

loicpoulain replied:
fixed

jfischer-phytec-iot commented on drivers/video/video_mcux_elcdif.c:
```diff
@@ -0,0 +1,425 @@
+	.set_signal = video_mcux_elcdif_set_signal,
+#endif
+#ifdef CONFIG_VIDEO_DISPLAY
+	.display_drv_api.blanking_on = display_video_blanking_on,
```
I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver, then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility". Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.

loicpoulain replied:
> I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver

Well, The video API is more generic than display, it's any device that can generate or consume video buffers (camera, encoder, display...), this API mainly allows to create zero-copy video pipelines. ELCDIF is not really a display but an 'LCD interface device' which can be connected to a 'LCD display device' and thus can be implemented as a generic video driver (output video device).

> then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility".

Exactly, this avoids code duplication. Moreover display is currently requested because LVGL relies on Display API and cannot interact directly with Video API.

> Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.

Not for now, but I need to discuss it with @vanwinkeljan to check if it's possible to converge at some point. Display API allows to write specific region of a frame buffer, video only takes full frame buffers pointers.

jfischer-phytec-iot replied:
> > I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver
> 
> Well, The video API is more generic than display, it's any device that can generate or consume video buffers (camera, encoder, display...), this API mainly allows to create zero-copy video pipelines. ELCDIF is not really a display but an 'LCD interface device' which can be connected to a 'LCD display device' and thus can be implemented as a generic video driver (output video device).


No, it is not and definitely not suitable for displays. ELCDIF is a generic display controller, and what we usually have are display controller drivers in `drivers/display`. ELCDIF driver should stay there.

> 
> > then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility".
> 
> Exactly, this avoids code duplication. Moreover display is currently requested because LVGL relies on Display API and cannot interact directly with Video API.


I do not see any duplication if you would use display API.

> 
> > Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.
> 
> Not for now, but I need to discuss it with @vanwinkeljan to check if it's possible to converge at some point. Display API allows to write specific region of a frame buffer, video only takes full frame buffers pointers.

Display API does not forbid you to write a full frame at once. 
I am not entire happy with the display API but I am sure it should not be mixed with video API. I will block this PR. APIs should be kept simple and do what they were created for.

loicpoulain replied:
> > > I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver
> > 
> > 
> > Well, The video API is more generic than display, it's any device that can generate or consume video buffers (camera, encoder, display...), this API mainly allows to create zero-copy video pipelines. ELCDIF is not really a display but an 'LCD interface device' which can be connected to a 'LCD display device' and thus can be implemented as a generic video driver (output video device).
> 
> No, it is not and definitely not suitable for displays. ELCDIF is a generic display controller, and what we usually have are display controller drivers in `drivers/display`. ELCDIF driver should stay there.
> 
> > > then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility".
> > 
> > 
> > Exactly, this avoids code duplication. Moreover display is currently requested because LVGL relies on Display API and cannot interact directly with Video API.
> 
> I do not see any duplication if you would use display API.
> 
> > > Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.
> > 
> > 
> > Not for now, but I need to discuss it with @vanwinkeljan to check if it's possible to converge at some point. Display API allows to write specific region of a frame buffer, video only takes full frame buffers pointers.
> 
> Display API does not forbid you to write a full frame at once.
> I am not entire happy with the display API but I am sure it should not be mixed with video API. I will block this PR. APIs should be kept simple and do what they were created for.

I do not understand, I do not modify any API here, I moved the ELCDIF as a simpler video driver which can implement either video API or display API.

jfischer-phytec-iot replied:
> I do not understand, I do not modify any API here, I moved the ELCDIF as a simpler video driver which can implement either video API or display API. The current display API allows to write frame (which implies copy), that's true but not to just display frame (zero copy). Video API is just a device type agnostic layer allowing to queue dequeue video buffer from/to any kind of video device (including display). If it's jsut about folder we can keep it under drivers/display/.

Do not move ELCDIF driver to video, do not use video API as display API. Implement a demuxer to pass the data from a video source to a device like display controller.

loicpoulain replied:
> > I do not understand, I do not modify any API here, I moved the ELCDIF as a simpler video driver which can implement either video API or display API. The current display API allows to write frame (which implies copy), that's true but not to just display frame (zero copy). Video API is just a device type agnostic layer allowing to queue dequeue video buffer from/to any kind of video device (including display). If it's jsut about folder we can keep it under drivers/display/.
> 
> Do not move ELCDIF driver to video, do not use video API as display API. Implement a demuxer to pass the data from a video source to a device like display controller.

This is something to discuss, since I tend to disagree, if requested I'll create a separated video driver for ELCDIF, indeed this is a perfect candidate for video interface API since it just pushes data to a display and thus can be abstracted by an output device. Display devices can be quite complicated and controlled via an additional interface like I2C to configure/retrieve resolution, format... I don't see why adding a video to display 'demuxer' would be better than the display to video one as proposed in this PR (video is more generic than display). I just want to keep this simple and efficient without breaking the current display API/code, not make it just 'work' via overkill layers.

jfischer-phytec-iot replied:
> This is something to discuss, since I tend to disagree, if requested I'll create a separated video driver for ELCDIF, indeed this is a perfect candidate for video interface API since it just pushes data to a display and thus can be abstracted by an output device.

Again, do not move the display controller driver to video, use display API for display controller please. We usually push the data to the displays and use display API. 

>Display devices can be quite complicated and controlled via an additional interface like I2C to configure/retrieve resolution, format... 

Video devices too, so what?

>I don't see why adding a video to display 'demuxer' would be better than the display to video one as proposed in this PR (video is more generic than display). I just want to keep this simple and efficient without breaking the current display API/code, not make it just 'work' via overkill layers.

You do not have to break anything, just read the video data and write it to the display controller using display API. With the proper adaption layer you can use any display controller in the tree. Like you used socket API in the tcpserversink sample and did not tried to hack a video API - mcux ethernet API video driver.

loicpoulain replied:
> > This is something to discuss, since I tend to disagree, if requested I'll create a separated video driver for ELCDIF, indeed this is a perfect candidate for video interface API since it just pushes data to a display and thus can be abstracted by an output device.
> 
> Again, do not move the display controller driver to video, use display API for display controller please. We usually push the data to the displays and use display API.

> > I don't see why adding a video to display 'demuxer' would be better than the display to video one as proposed in this PR (video is more generic than display). I just want to keep this simple and efficient without breaking the current display API/code, not make it just 'work' via overkill layers.
> 
> You do not have to break anything, just read the video data and write it to the display controller using display API. With the proper adaption layer you can use any display controller in the tree. Like you used socket API in the tcpserversink sample and did not tried to hack a video API - mcux ethernet API video driver.

It is indeed possible, but it's a poor solution, video API abstracts the components (camera, encoder, display) in a unified way, allowing user to use a simple common interface to retrieve and align/configure formats as well as to deal with buffers (cf capture_and_display sample). In that case, using display adds unnecessary complexity (different structures, functions...), much bigger memory footprint (display has two private dedicated buffers), latency/cpu-cycles (display implies copying buffer)... Moreover the current proposal is more flexible since display interface (ELCDIF) and display (LCD) are two separated devices, allowing to 'add/connect' any device to the interface without modifying the interface driver itself (dedicated LCD driver). That why I push for this unified solution. I can not force anything, but I don't get why it would not be suitable for zephyr to have such support, knowing that it still backward compatible with Display API.

vanwinkeljan replied:
> much bigger memory footprint

This is just due to the fact you are supporting partial buffer writes in case you would enforce writing the whole buffer at once this buffers are not needed. Alternatively you could deque all video buffers and write the data to them and enqueue them.

> latency/cpu-cycles

This is the price you pay for using/supporting `display_write` API, the display API also provides an option to get the framebuffer directly `display_get_framebuffer` and perform actions directly on the framebuffer. Actually it is a bit strange that the ELCDIF display driver never implemented this support but I think this was more a time constraint then a technical issue. (Possibly we need to extend the display API with `display_put_framebuffer`).

> Moreover the current proposal is more flexible since display interface (ELCDIF) and display (LCD) are two separated devices

I like this approach but this has nothing to do with video API and could have been achieved with the display API.

@jfischer-phytec-iot I have no problem in deprecating the display API and go to one unified API. But it is a hard requirement to support partial buffers and the unified API should have minimal dependencies to other subsystems.

loicpoulain replied:
> > much bigger memory footprint
> 
> This is just due to the fact you are supporting partial buffer writes in case you would enforce writing the whole buffer at once this buffers are not needed. Alternatively you could deque all video buffers and write the data to them and enqueue them.
> 
> > latency/cpu-cycles
> 
> This is the price you pay for using/supporting `display_write` API, the display API also provides an option to get the framebuffer directly `display_get_framebuffer` and perform actions directly on the framebuffer. Actually it is a bit strange that the ELCDIF display driver never implemented this support but I think this was more a time constraint then a technical issue. (Possibly we need to extend the display API with `display_put_framebuffer`).

Yes, since we need a way to dequeue/enqueue (get/put). An other point is that getting buffer is a blocking function in display API, so it could request its own dedicated thread to perform concurrent operations (getting camera buffer, sw image processing).

> > Moreover the current proposal is more flexible since display interface (ELCDIF) and display (LCD) are two separated devices
> 
> I like this approach but this has nothing to do with video API and could have been achieved with the display API.

Correct.

> @jfischer-phytec-iot I have no problem in deprecating the display API and go to one unified API. But it is a hard requirement to support partial buffers and the unified API should have minimal dependencies to other subsystems.

well, that why I went with this backward compatible solution, display API can not really be deprecated for now, but I would be happy to work with you on unifying the APIs (probably progressively, adding specific video controls for display, LVGL support, etc), maybe starting by using common formats, pixel code, buffer formats... Currently, partial write is possible with video API since write can be done directly in the buffer before enqueuing it for output. At the end I agree it's is possible to improve the display API, but I'm much in favor of creating one unified API and make things easier for user (like it can be with Linux Video APIs, v4l2, gstreamer...).

jfischer-phytec-iot replied:
So really slow down. You do not have a solution here, it is a bad hack nothing more. Even with an one common overloaded API you have to implement a system between data source and data drain because they would not always have the same format like resolution and pixel format. Your video API is far from the point to be proven even for video sources, I can only see the CSI interface and one driver for the sensors. Display API is not perfect but much more mature. Again, the ELCDIF is a display controller and the driver should use display API. If you want to use ELCDIF or an other display controller as a drain for the data from a video source then add a sample/subsystem/layer for that, like it was done for tcpserversink!

loicpoulain replied:
I understand the maturity point, though I don't want to replace the display API here, but make the ELCDIF support both display API and video API, which normally does not cause any regression and is, in my opinion, a good compromise. This makes the sample easier to understand, format configuration/alignment straightforward and tested with compatible devices (well, just the ELCDIF + RK043FN02H for now) since I'm not sure any display can support 'video'. I can take a step back and go with the display API solution which would certainly work. But then I don't know what to do after that, should we create a dedicate API for camera, encoder, decoder, display, video output... or trying to move on a unified version. Don't want an endless ping-pong here, I know we don't agree on the way to go, just trying to find a way to move forward on this PR and on next steps.

dbkinder reviewed (approved):
+1 for doc changes, thanks!

Commit:
drivers: video: Add NXP MCUX ELCDIF video driver

Add driver for eLCDIF interface, relying on fsl_elcdif.
This controller is a LCD controller capable of dma transfer.
The eLCDIF video device needs to be coupled with a LCD video
device, which can be specified via the lcd-label property.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
drivers: display: Add display to video shim driver

A video device is a generic piece that can generate or consume
video data. This patch allows any video output device to act as
a display, by implementing a generic display to video shim layer.

This allows a video output device to inherit from display API
using the shim layer helpers.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
drivers: display: mcux_elcid: migrate to video

eLCDIF is able to act as a standard video output device, allowing
to create zero-copy end-to-end video pipeline. Keep compatibility
with display API using inheritance and display-to-video shim layer.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

microbuilder commented:
Tested on the i.mxrt1064-evk and works for me, and happy to approve if the CI issues for the PR are resolved.

Commit:
modules: Update nxp hal for CI

Point to nxp hal PR21 (DO NOT MERGE)
https://github.com/zephyrproject-rtos/hal_nxp/pull/21

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
video: mcux_csi: Fix undefined sensor_label prop

If dts does not define the sensor_lable prop, don't use it.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
dts: bindings: display: rk043fn02h-ct

Add required label property.
This property is unconditionnaly used by the driver.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
> Tested on the i.mxrt1064-evk and works for me, and happy to approve if the CI issues for the PR are resolved.

Done!

jfischer-no reviewed (changes requested):
I do not accept/agree with the changes in this PR at all. 
                                                                              
The first commit adds a display controller driver using video API instead of improve the driver in the tree and using display controller API we have.

capture_and_display sample is not generic although it pretends to be, it can not, but it might if it would use display API for the display.

This PR introduces unnecessary complexity to display API, unnecessary (shim) drivers and configuration. It justifies the changes only on the basis of some samples made or tested only for a specific hardware.                                       
                                                                                                                                              
The display controller driver should use display controller API.                                                                              
We do not need unnecessary complexity in the APIs and the drivers.                                                                            
How the common attributes like pixel format can be used by both APIs is another story.

loicpoulain commented:
Maybe I should rename the sample then, since it does not use the 'display' API. This is an attempt to make things more generic and have a common interface for all devices that are able to handle video stream (buffers). The imx device being capable of that I simply make it a video device, and by simple inheritance, a display device, it does no break anything and would allow any future similar device to be compatible with display API for free. That's permit zero-copy video pipeline, and is much more simple to use as the API is unified in such cases. In short it does not break anything, add a feature, an optimized video sample, and keep display API untouched.

Improving display API to make it more or less compatible with video is possible, but this is no the point, video is a generic representation of any video device, display is a layer above. The sample name (video_and_display) is probably misleading and could be renamed video pipeline or so...

> I do not accept/agree with the changes in this PR at all.
> 

> 
> capture_and_display sample is not generic although it pretends to be, it can not, but it might if it would use display API for the display.
> 
> This PR introduces unnecessary complexity to display API, unnecessary (shim) drivers and configuration. It justifies the changes only on the basis of some samples made or tested only for a specific hardware.
> 
> The display controller driver should use display controller API.
> We do not need unnecessary complexity in the APIs and the drivers.
> How the common attributes like pixel format can be used by both APIs is another story.

pabigot commented on drivers/video/video_mcux_csi.c:
```diff
@@ -404,7 +404,9 @@ static const struct video_driver_api video_mcux_csi_driver_api = {
 static const struct video_mcux_csi_config video_mcux_csi_config_0 = {
 	.base = (CSI_Type *)DT_INST_REG_ADDR(0),
+#if DT_INST_NODE_HAS_PROP(0, sensor_label)
 	.sensor_label = DT_INST_PROP(0, sensor_label),
```
This seems unrelated to this feature, and should perhaps be part of a PR that changes the property from a label to a phandle.  For the purposes of this PR it's OK to keep it this way.

loicpoulain replied:
I've added the change in this PR

pabigot commented on dts/bindings/display/rocktech,rk043fn02h-ct.yaml:
```diff
@@ -6,3 +6,7 @@ description: Rocktech LCD module with LED backlight and capacitive touch panel
+
+properties:
+    label:
+      required: true
```
Is this still required if the relationship is managed through phandle references rather than the label?

loicpoulain replied:
Yes we can remove that when switching to phandle.

pabigot reviewed (changes requested):
A couple comments on the devicetree aspects.  We will be using devicetree node relationships to support power management which means using node references rather than text labels to indicate relationships between nodes.

I see there is existing practice on this board to use a label rather than a phandle to associate one node with another but that should be avoided (and fixed for the sensor case, though not in this PR).

pabigot commented:
My perceptions from the presentation and discussion in the API telecon:

It makes sense to me that a display can be a sink in a video pipeline, and as such would be a video device.  For a display driver to serve as a sink with equivalent performance it would need two features:
* Synchronization so that the frames it is to display can be handed back to the video pipeline for re-use when the display driver finishes consuming the data (copying to its frame buffer); or
* An API to change the underlying frame buffer that the display uses (supporting zero-copy).

The current implementation in this PR of providing a shim that makes the video sink look like a display sink makes sense to me.  Future enhancements to the display could allow use of it as a sink directly, but it doesn't seem entirely reasonable to block adoption of this solution until that's done.

At this time I would approve this once the devicetree issue is addressed.

pabigot commented on drivers/display/display_dummy.c:
```diff
@@ -24,50 +24,50 @@ static int dummy_display_init(struct device *dev)
 }
 
-static int dummy_display_write(const struct device *dev, const uint16_t x,
+static int dummy_display_write(struct device *dev, const uint16_t x,
```
This change is also problematic as an API change.  I tried removing the commit and the sample still builds.

There's work in progress to make all `struct device` instances const qualified "soon", so this change would be reverted when that happens.  It'd be better to not do this if you can avoid it.  Can you eliminate the warnings by adding casts inside the shim implementation (if you need a non-const driver reference)?

loicpoulain replied:
done

Commit:
fixup fsl imx dts bindings

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented on drivers/video/video_mcux_elcdif.c:
```diff
@@ -0,0 +1,413 @@
+static const struct video_mcux_elcdif_config video_mcux_elcdif_config_0 = {
+	.base = (LCDIF_Type *)DT_INST_REG_ADDR(0),
+#if DT_INST_NODE_HAS_PROP(0, lcd_label)
+	.lcd_label = DT_INST_PROP(0, lcd_label),
```
yes, used DT_LABEL(DT_INST_PHANDLE(0, lcd)

loicpoulain commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -99,6 +101,7 @@ arduino_i2c: &lpi2c1 {};
 
 &lcdif {
 	status = "okay";
+	lcd-label = "rk043fn02h-ct";
```
done

loicpoulain commented on dts/bindings/display/fsl,imx6sx-lcdif.yaml:
```diff
@@ -16,3 +16,8 @@ properties:
       required: true
+
+    lcd-label:
+      type: string
```
done

pabigot reviewed (changes requested):
Premature approval: changes are good, but the fixup commit needs to be squashed into the bindings commit.

jfischer-phytec-iot commented on samples/video/capture_and_display/README.rst:
```diff
@@ -0,0 +1,66 @@
+Description
+***********
+
+This sample application uses the Video API to retrieve video frames from the
```
This sample is not generic. It can only be used with LCDIF on IMXRTxxxx. But it could be almost generic if it would use display driver API and configure VIDEO_SW_GENERATOR to output compatible frames (example https://github.com/jfischer-phytec-iot/zephyr/blob/test-video2display/samples/video/video2display/src/main.c). Almost, because the VIDEO_SW_GENERATOR is (again) half-baked and does not support different pixel formats.

Additional:  this sample (and `samples/video/capture`) should be in **samples/drivers** because it is a sample how to use video driver API.

loicpoulain replied:
> But it could be almost generic if it would use display driver API and configure VIDEO_SW_GENERATOR to output compatible frames

I'm not against such a sample that use the display API, it was just not optimized and synchronized enough for the 'zero-copy' use case.

> because the VIDEO_SW_GENERATOR is (again) half-baked and does not support different pixel formats.

What do you mean, and in which way the display API solves that issue?

jfischer-phytec-iot commented on samples/video/capture_and_display/README.rst:
```diff
@@ -0,0 +1,66 @@
+    Capture started
+
+Once capture has started, the display shows camera captured frames. For
+:ref:`mimxrt1064_evk` board it could be necessary to hard reset the board (SW3
```
Why is it necessary to reset the board? VIDEO_SW_GENERATOR works fine with display driver API without additional reset.

loicpoulain commented on drivers/video/video_mcux_elcdif.c:
```diff
@@ -0,0 +1,413 @@
+			 kELCDIF_VsyncActiveLow |
+			 kELCDIF_HsyncActiveLow |
+			 kELCDIF_DriveDataOnRisingClkEdge;
+	data->elcdif_config.dataBus = kELCDIF_DataBus16Bit;
```
Right, this is indeed an improvement that can be added.

loicpoulain commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,178 @@
+
+int display_video_blanking_off(const struct device *dev)
+{
+	return -ENOTSUP;
```
The display_mcux_elcdif driver was not implementing that, so I've no way to test it. But this is something that can be added via a video-ctrl (video_set_ctrl()) call.

loicpoulain commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,178 @@
+	video_get_format((struct device *)dev, VIDEO_EP_IN, &fmt);
+
+	switch (pixel_format) {
+	case PIXEL_FORMAT_RGB_565:
```
Well, that's true, this is something to add if used with different pixel format.

loicpoulain commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,178 @@
+	bsize = fmt.height * fmt.pitch;
+
+	/* Alloc two buffers (double buffering) and enqueue */
+	vbufa = video_buffer_alloc(bsize);
```
Right, for now using the shim layer does not do better than pure display driver, it's just a way to make a video device (also) compatible with display API.

If the device is not able to use main memory for buffer, for sure a copy is requested, and in that case, video and display implementation would be equivalent. That would however allow to use the unified video API, providing a unified interface for configuring 'video' elements and exchanging buffers, without having to deal with multiples APIs.

loicpoulain commented on drivers/display/display_video.c:
```diff
@@ -0,0 +1,178 @@
+
+	/* LVGL no double buffering aware support ? just copy previous one */
+	src = (vbuf == vbufa) ? vbufb->buffer : vbufa->buffer;
+	memcpy(vbuf->buffer, src, vbuf->bytesused);
```
well only the video API is zero-copy approach for now. But good to now about that option

loicpoulain commented on boards/arm/mimxrt1064_evk/pinmux.c:
```diff
@@ -130,7 +130,7 @@ static int mimxrt1064_evk_init(struct device *dev)
 #endif
 
-#if DT_NODE_HAS_STATUS(DT_NODELABEL(lcdif), okay) && CONFIG_DISPLAY
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(lcdif), okay) && CONFIG_VIDEO_MCUX_ELCDIF
```
> Used as display controller driver, it has the same memory usage as display_mcux_elcdif.c but unnecessary complexity over shim. zero-copy has nothing to do with display use case at all, and shim driver uses memcpy explicitly.

I do not improve in any way the usage of mcux elcdif via the display API, that's correct, but that's not the goal. The goal being to expose the mcux elcif as video output device, to use with the video API, which is in that case the optimized path. The display shim layer being here just for 'backward' compatibility.

So, given your feedback, and in order to move forward, maybe it's better to not create that shim layer and to live with two different drivers for now. Then check what is possible to do to either improve the display API and make both APIs more 'compatible' (unify pixel formats, buffer descriptors...) or even merge both if technically acceptable.
